### PR TITLE
[WIP] Configuration overhaul

### PR DESCRIPTION
**Pull request by djberg96**
Do not merge yet. Work in progress.

After hitting an issue with using service classes within other service classes, I decided that it would be better to revamp the configuration handling as a starting point.

Part of that means splitting out the configuration handler into its own class and in its own file. I think this will make maintenance easier than having a struct + block with methods defined on the fly, since it's easier to identify what is happening and where.

Another advantage is that we can more robustly set default values, instead of creating a configuration object, and then setting values after the fact.

TL;DR: I think splitting it out into its own bonafide class will make it easier to maintain, test and develop going forward.

### CODE REVIEW

File: lib/azure/armrest/configuration.rb [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        tokens[key]
+      end
+
+      # Set the value of the tokens class variable hash.
+      def []=(key, value)
+        tokens[key] = value
+      end
+
+      # Returns the tokens class variable hash.
+      def tokens
```
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: We should not expose `tokens` as it is used to manage internal generated tokens (excluding those directly set by user). Same reason `[](key)` and `[]=(key, value)` are not needed.

Comment (djberg96) [inline lib/azure/armrest/configuration.rb]: Ok, though we're effectively exposing it by making it a class variable. I was hoping to eventually dispense with that and make it a regular instance variable.

File: lib/azure/armrest/configuration.rb [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+      attr_accessor :content_type
+
+      # The accept type specified for http request results. The default is 'application/json'
+      attr_accessor :accept
+
+      # The bearer token passed to all http requests.
+      attr_writer :token
+
+      # Set the time in which the token expires.
+      attr_writer :token_expiration
```
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: combine them into one method like `def set_token(token, expiration)` as they need to be set externally at the same time

File: lib/azure/armrest/configuration.rb [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+      def as_cache_key
+        "#{grant_type}_#{tenant_id}_#{client_id}_#{client_key}"
+      end
+
+      alias cache_key as_cache_key
+
+      # Returns the token for the given cache_key, or sets it if it does not
+      # exist or it has expired.
+      #
+      def token
```
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: This implementation is buggy which causes the spec test failures. 
```
      def token
        @token, @token_expiration = tokens[cache_key] if @token.nil?

        if @token.nil? || Time.now > @token_expiration
          @token, @token_expiration = fetch_token
        end

        @token
      end
```

File: lib/azure/armrest/configuration.rb (hunk 2 above)
Comment (djberg96) [inline lib/azure/armrest/configuration.rb]: Agreed.

File: lib/azure/armrest/configuration.rb (hunk 3 above)
Comment (djberg96) [inline lib/azure/armrest/configuration.rb]: Yep, thanks.

**Comment by djberg96** [comment]
@bzwei For now I've redefined a token= method in order to pass the specs. It will internally call the set_token method with a default token expiration.

File: lib/azure/armrest/configuration.rb [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+      # time, use the :set_token method instead.
+      #
+      def token=(value)
+        set_token(value)
+      end
+
+      # Set the token value and expiration time. If no time is specified, then
+      # the time is set for 1 hour.
+      #
+      def set_token(token, token_expiration = Time.now + 3600)
```
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: It is Azure that configures the token expiration. Setting a default expiration has no meaning and misleads.

File: lib/azure/armrest/configuration.rb [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+        # Set defaults if not already set
+        @accept       ||= 'application/json'
+        @content_type ||= 'application/json'
+        @grant_type   ||= 'client_credentials'
+        @api_version  ||= '2015-01-01'
+      end
+
+      # Returns the tokens class variable hash.
+      def tokens
```
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: We really don't need this method as the hash has no value to user. Exposing it actually incurs dangers of being unintentionally overwritten. Besides this kind of method should be a class method.

Comment (djberg96) [inline lib/azure/armrest/configuration.rb]: Agreed, I meant to remove that.

File: lib/azure/armrest/configuration.rb [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+      alias cache_key as_cache_key
+
+      # Returns the token for the given cache_key, or sets it if it does not
+      # exist or it has expired.
+      #
+      def token
+        @token, @token_expiration = @@tokens[cache_key] if @token.nil?
+
+        if @token.nil? || Time.now > (@token_expiration || Time.new(0))
```
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: I am working on another PR where I got a Rubocop warning: Do not use Time.now without zone. Use one of Time.zone.now, Time.now.current, Time.now.in_time_zone, Time.now.utc, Time.now.getlocal, Time.now.iso8601, Time.now.jisx0301, Time.now.rfc3339, Time.now.to_i, Time.now.to_f instead.

I thus recommend to use integer for expiration, thus second part of the condition can be `Time.now.to_i > (@token_expiration || 0)`

File: lib/azure/armrest/configuration.rb [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+          token_url,
+          :grant_type    => grant_type,
+          :client_id     => client_id,
+          :client_secret => client_key,
+          :resource      => Azure::Armrest::RESOURCE
+        ))
+
+        token = 'Bearer ' + response['access_token']
+
+        @@tokens[cache_key] = [token, Time.now + response['expires_in'].to_i]
```
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: This becomes `[token, Time.now.to_i + response['expiries_in'].to_i]`

File: lib/azure/armrest/armrest_service.rb [hunk 8]
```diff
@@ -102,23 +57,10 @@ def fetch_token
       # the new portal or the New-AzureRoleAssignment powershell command.
       #
       def self.configure(options)
```
Comment (bzwei) [inline lib/azure/armrest/armrest_service.rb]: The old design place the configuration class inside `ArmrestService`, thus make sense to provide this factory method to instantiate a configuration object.

The new design move the configuration class as an independent class. A normal usage will be for a user to directly instantiate a configuration object and pass it to create services. This factory method should remain here only for backward compatible reason. 

However this factory method finds the `subscription_id` if not exist. This makes the factory not obsolete but necessary, which kind of fail the purpose of this refactoring work.

File: lib/azure/armrest/configuration.rb (hunk 6 above)
Comment (djberg96) [inline lib/azure/armrest/configuration.rb]: Are those rails methods? I'm afraid I don't agree with that warning for this project.

File: lib/azure/armrest/configuration.rb (hunk 7 above)
Comment (djberg96) [inline lib/azure/armrest/configuration.rb]: Converting these means we would also have to explicitly type check the :token_expiration argument, too.

File: lib/azure/armrest/armrest_service.rb (hunk 8 above)
Comment (djberg96) [inline lib/azure/armrest/armrest_service.rb]: I can move that into the configuration class.

File: lib/azure/armrest/configuration.rb (hunk 6 above)
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: Since expiration can be set from outside, we'd better leave it as type `Date`. However the comparison here should change to `Time.now.utc > (@token_expiration.utc || Time.new(0))`. 

Other improvements can be to initialize `@token_expiration` to `Time.new(0)` in the constructor, and in the setter of `@token_expiration` to convert the input date to correct zone.

Comment (djberg96) [inline lib/azure/armrest/configuration.rb]: Ok, I've made your suggested changes.

File: lib/azure/armrest/configuration.rb [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+      def fetch_token
+        token_url = File.join(Azure::Armrest::AUTHORITY, tenant_id, 'oauth2/token')
+
+        response = JSON.parse(ArmrestService.rest_post(
+          token_url,
+          :grant_type    => grant_type,
+          :client_id     => client_id,
+          :client_secret => client_key,
+          :resource      => Azure::Armrest::RESOURCE
+        ))
```
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: `))`?

Comment (djberg96) [inline lib/azure/armrest/configuration.rb]: @bzwei Hm? Are you saying you don't like the formatting? That was just a copy/paste from the source, actually. I can change it, though.

File: lib/azure/armrest/configuration.rb [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+
+        token = 'Bearer ' + response['access_token']
+
+        @@tokens[cache_key] = [token, Time.now.utc + response['expires_in'].to_i]
+      end
+
+      def fetch_subscription_id
+        return @subscriptions[cache_key] if @subscriptions.has_key?(cache_key)
```
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: Prefer `key?(cache_key)`. Check with @matthewd

File: lib/azure/armrest/configuration.rb [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+      def set_token(token, token_expiration)
+        @token, @token_expiration = token, token_expiration
+      end
+
+      # Returns the expiration datetime for given key, or the current
+      # cache_key if no key is specified. If the value is nil, return
+      # the epoch time.
+      #
+      def token_expiration(key = nil)
+        key ? @@tokens[key].last : @@tokens[cache_key].last
```
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: how about
```
  key ||= cache_key
  @@tokens[key]
```

File: lib/azure/armrest/configuration.rb [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+
+      alias cache_key as_cache_key
+
+      # Returns the token for the given cache_key, or sets it if it does not
+      # exist or it has expired.
+      #
+      def token
+        @token, @token_expiration = @@tokens[cache_key] if @token.nil?
+
+        if @token.nil? || Time.now.utc > @token_expiration.utc
```
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: since the setter for `@token_expiration` already forces utc, no need `@token_expiration.utc` here.

File: lib/azure/armrest/configuration.rb [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+
+        # Set defaults if not already set
+        @accept           ||= 'application/json'
+        @content_type     ||= 'application/json'
+        @grant_type       ||= 'client_credentials'
+        @api_version      ||= '2015-01-01'
+        @token_expiration ||= Time.new(0)
+
+        # Used internally
+        @subscriptions = {}
```
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: I don't see why `@subscriptions` is needed. It stores only one entry. In the old code it is a class variable. 
Consider remove `@subscriptions` or change it to class variable.

File: lib/azure/armrest/armrest_service.rb [hunk 14]
```diff
@@ -102,52 +55,15 @@ def fetch_token
...
-          :authorization => config.token
-        )
-
-        hash = JSON.parse(response)["value"].first
-
-        raise ArgumentError, "No associated subscription found" if hash.empty?
-
-        id = hash.fetch("subscriptionId")
-        @@subscriptions[config.as_cache_key] = id
+        Azure::Armrest::Configuration.new(options)
```
Comment (bzwei) [inline lib/azure/armrest/armrest_service.rb]: Should we document this factory method obsolete and update your wiki to guide users to directly create a configuration object?

File: lib/azure/armrest/configuration.rb (hunk 9 above)
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: Sorry. I miscounted the paring of parentheses. Never mind.

File: lib/azure/armrest/configuration.rb (hunk 11 above)
Comment (djberg96) [inline lib/azure/armrest/configuration.rb]: Yep. We can set it as the default value in the method, actually.

File: lib/azure/armrest/configuration.rb (hunk 13 above)
Comment (djberg96) [inline lib/azure/armrest/configuration.rb]: In theory you could have multiple subscriptions per configuration.

Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: yes. but in your `fetch_subscription_id` only takes the first subscription, and it is referenced by `cache_key` which is unique per configuration.

Comment (djberg96) [inline lib/azure/armrest/configuration.rb]: Right, the subscriptions instance var should probably be changed to Hash.new([]). I'm trying to decide how the interface should work.

File: lib/azure/armrest/configuration.rb [hunk 15]
```diff
@@ -0,0 +1,183 @@
+module Azure
+  module Armrest
+    class Configuration
+      # Used to store unique token information.
+      @@tokens = Hash.new([])
```
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: should be `@@tokens = {}`

File: lib/azure/armrest/configuration.rb [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+      # Example:
+      #
+      #   config = Azure::Armrest::Configuration.new(
+      #     :client_id  => 'xxxx',
+      #     :client_key => 'yyyy',
+      #     :tenant_id  => 'zzzz'
+      #   )
+      #
+      def initialize(hash)
+        @subscriptions = Hash.new([]) # Used internally
```
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: `@subscriptions = {}`

Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: should be class variable `@@subscriptions`, referenced by `cache_key`, unique per configuration.

Comment (djberg96) [inline lib/azure/armrest/configuration.rb]: I don't agree. Since it's unique per configuration, it does not require a class variable. That's one of the changes I wanted to make.

File: lib/azure/armrest/configuration.rb [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+
+        url = File.join(Azure::Armrest::RESOURCE, "subscriptions?api-version=#{api_version}")
+
+        response = ArmrestService.rest_get(
+          url,
+          :content_type  => content_type,
+          :authorization => token
+        )
+
+        hash = JSON.parse(response)['value'].first
```
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: Does `JSON.parse(response)['value']` gives you all subscriptions? then all of them should be saved. Otherwise I don't see why we need to maintain a list of subscriptions per configuration.

File: lib/azure/armrest/configuration.rb (hunk 16 above)
Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: The whole implementation you only use one key for the hash, which is `cache_key`.

File: lib/azure/armrest/configuration.rb (hunk 17 above)
Comment (djberg96) [inline lib/azure/armrest/configuration.rb]: Yes, it does.

File: lib/azure/armrest/configuration.rb (hunk 16 above)
Comment (djberg96) [inline lib/azure/armrest/configuration.rb]: Do we ever modify a configuration object after it has been created? If not, I'm wondering why we need a cache_key at all.

Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: User can create multiple configuration objects from the same set of ids, but they shall share the same token, thus a key is needed. Alternatively you may want to override `hash` method, but be aware a hashcode is used for more than this purpose.

Comment (djberg96) [inline lib/azure/armrest/configuration.rb]: Only because we're using a class variable to store @@tokens, right? This made sense before we explicitly passed configuration objects to Service class initializers, but I don't know how much sense it makes now. Why not just have the Service instance get the token from the configuration object that was passed to it?

Maybe I need some code samples to see the situation you're describing.

Comment (bzwei) [inline lib/azure/armrest/configuration.rb]:
```
attrs = {...}
cf1 = Configuration.new(attrs)
cf2 = Configuration.new(attrs)
```
How do we ensure `cf1.token == cf2.token` is `true`? Through `@@tokens[cache_key]`

Comment (djberg96) [inline lib/azure/armrest/configuration.rb]: Why do we care if cf1.token == cf2.token in practice, though, if we're passing config objects directly to construct Service objects? Given:

    vms = Azure::Armrest::VirtualMachineService.new(cf1)

The vms object is only going to get its info from cf1, since it has everything it needs there, right?

Comment (bzwei) [inline lib/azure/armrest/configuration.rb]: We don't want to create a new token for each duplicated configuration. Just like if you already login to an account in a browser, you can continue to access the account in a new tab without login.

Comment (djberg96) [inline lib/azure/armrest/configuration.rb]: I don't know that I agree with that. I think that is something for end users to implement on their end if they want.

But, I'll continue down that path for now. Since you brought up config equality, I'm thinking this class should implement Comparable.
